lib/ret/media_resolver.ex

```diff
@@ -45,6 +48,40 @@ defmodule Ret.MediaResolver do
     resolve_non_video(query, root_host)
   end
 
+  # For youtube.com, we need to rate limit requests. Only do one at a time on this host.
+  # Also compute ttl here based upon expire, to localize youtube.com specific logic.
+  def resolve(%MediaResolverQuery{} = query, "youtube.com" = root_host) do
+    Mutex.under(MediaResolverMutex, :youtube, @youtube_rate_limit_timeout, fn ->
+      now = System.system_time(:millisecond)
+      res = resolve_with_ytdl(query, root_host, query |> ytdl_format(root_host))
+
+      {:commit, %Ret.ResolvedMedia{uri: %URI{query: youtube_query}} = resolved_media} = res
+
+      # YouTube returns a 'expire' which has timestamp of expiration.
+      resolved_media =
+        with parsed_youtube_query <- URI.decode_query(youtube_query),
+             expire when is_binary(expire) <- Map.get(parsed_youtube_query, "expire"),
+             {expire_s, _} <- Integer.parse(expire),
+             # Don't use custom TTL if expires in < 2 minutes
+             ttl_s when ttl_s > 120 <- expire_s - System.system_time(:second) do
+          # Expire a minute early
+          resolved_media |> Map.put(:ttl, ttl_s * 1000 - 60000)
+        else
+          _ -> resolved_media
+        end
+
+      request_duration_ms = System.system_time(:millisecond) - now
+      remaining_duration_ms = @youtube_ms_per_request_rate_limit - request_duration_ms
+
+      if remaining_duration_ms > 0 do
```

Comment:
Not fully following how the mutex API works, but doesn't this add delay to every request, even if no other request is actually pending? Seems like we would only want to wait on getting the mutex in a subsequent request, but not delay the first request that came in. (basically the piece I am missing here is when does the outer block actually return a result? Is it before or after this sleep?

Reply:
it'd be after the sleep, so good point. i think to do what you want we need to grab the mutex, do the request, and then spawn a process in the critical section to free the mutex after 4 seconds (and return immediately once the work is done.) A little hairy, but much better than stalling the requests.